core/src/test/java/com/linecorp/armeria/server/ServerListenerTest.java

```diff
@@ -100,4 +102,389 @@ public void serverStarting(Server server) throws Exception {
                 .hasCauseInstanceOf(IllegalStateException.class)
                 .hasRootCause(ex);
     }
+
+    @Test
+    void testGracefulShutdownPeriodicExecutorWaitIndefinitelyOnServerStopping() throws Exception {
+        final CompletableFuture<Void> stopFuture;
+        try (CloseableRunnable task = new CloseableRunnable() {
+            @Override
+            public void run() {
+                try {
+                    Thread.sleep(1000);
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                }
+            }
+        }) {
+            final ScheduledExecutorService periodic = Executors.newSingleThreadScheduledExecutor();
+            periodic.scheduleAtFixedRate(task, 0, 1000, TimeUnit.MILLISECONDS);
+            final ServerListener sl = ServerListener.builder()
+                                                    .shutdownWhenStopping(periodic, 0)
+                                                    .build();
+
+            final Server server = Server.builder()
+                                        .service("/", (req, ctx) -> HttpResponse.of("Hello!"))
+                                        .serverListener(sl)
+                                        .build();
+            server.start().get();
+            stopFuture = server.stop();
+            Thread.sleep(5000);
```

Comment:
Could you reduce these values so that the amount of total build time isn't increased too much?

Reply:
I combined the 12 divided tests into one to reduce the total test execution time. The tests, which used to take 60 seconds to complete, have now been reduced to 20 seconds.

Note: I attempted to terminate the ExecutorService instances concurrently without waiting for them to shut down sequentially. However, due to Armeria's ServerListener waiting for the synchronous termination of registered listeners, I had no choice but to wait for Executor termination one by one.